### PR TITLE
Fix behaviour of auto generated CRD KubeApis and KubeObjectStores

### LINKED ISSUE #7364: App crash on LENS /extension/aquasecurity--trivy-operator-lens-extension/vulnerabilityreports
On first TAB VulnerabilityReports

```
Error stack:
TypeError: e.getImageRef is not a function
    at d.render (/Users/mateuszjasinski/.k8slens/extensions/aquasecurity--trivy-operator-lens-extension/dist/renderer.js:1:20572)
    at qi (https://d59425e1be660b64edbcc0d8aaf4035f.lens.app:50318/build/lens.js?27361ea8c984dddf9250:566552:188)
    at pi (https://d59425e1be660b64edbcc0d8aaf4035f.lens.app:50318/build/lens.js?27361ea8c984dddf9250:566551:173)
    at ck (https://d59425e1be660b64edbcc0d8aaf4035f.lens.app:50318/build/lens.js?27361ea8c984dddf9250:566634:427)
    at bk (https://d59425e1be660b64edbcc0d8aaf4035f.lens.app:50318/build/lens.js?27361ea8c984dddf9250:566615:347)
    at ak (https://d59425e1be660b64edbcc0d8aaf4035f.lens.app:50318/build/lens.js?27361ea8c984dddf9250:566615:278)
    at Tj (https://d59425e1be660b64edbcc0d8aaf4035f.lens.app:50318/build/lens.js?27361ea8c984dddf9250:566615:138)
    at Lj (https://d59425e1be660b64edbcc0d8aaf4035f.lens.app:50318/build/lens.js?27361ea8c984dddf9250:566608:163)
    at https://d59425e1be660b64edbcc0d8aaf4035f.lens.app:50318/build/lens.js?27361ea8c984dddf9250:566488:115
    at exports.unstable_runWithPriority (https://d59425e1be660b64edbcc0d8aaf4035f.lens.app:50318/build/lens.js?27361ea8c984dddf9250:591052:343) 
``` 

**To Reproduce**
1. Click on any position in list.
2. See error

**Expected behavior**
See some information not ERROR.

**Environment (please complete the following information):**
- Lens Version: Lens: 2023.3.100736-latest
- OS:OSX 13.2.1
- Installation method: HELM

**Logs:**

```
status.reason":"Error","status.message":"2023-03-16T12:05:10.774Z\t\u001b[31mFATAL\u001b[0m\timage scan error: scan error: unable to initialize a scanner: unable to initialize a docker scanner: 4 errors occurred:\n\t* unable to inspect the image (xxxxxx): Cannot connect to the Docker daemon at unix:///var/run/docker.sock.
```

**Kubeconfig:**
Quite often the problems are caused by malformed kubeconfig which the application tries to load. Please share your kubeconfig, remember to remove any secret and sensitive information.
```
spec:
  replicas: 1
  selector:
    matchLabels:
      app.kubernetes.io/instance: trivy-operator
      app.kubernetes.io/name: trivy-operator
  template:
    metadata:
      creationTimestamp: null
      labels:
        app.kubernetes.io/instance: trivy-operator
        app.kubernetes.io/name: trivy-operator
    spec:
      containers:
        - name: trivy-operator
          image: ghcr.io/aquasecurity/trivy-operator:0.12.1-ubi8
          ports:
            - name: metrics
              containerPort: 8080
              protocol: TCP
            - name: probes
              containerPort: 9090
              protocol: TCP
          env:
            - name: OPERATOR_NAMESPACE
              value: trivy
            - name: OPERATOR_TARGET_NAMESPACES
            - name: OPERATOR_EXCLUDE_NAMESPACES
            - name: OPERATOR_TARGET_WORKLOADS
              value: >-
                pod,replicaset,replicationcontroller,statefulset,daemonset,cronjob,job
            - name: OPERATOR_SERVICE_ACCOUNT
              value: trivy-operator
            - name: OPERATOR_LOG_DEV_MODE
              value: 'false'
            - name: OPERATOR_SCAN_JOB_TTL
            - name: OPERATOR_SCAN_JOB_TIMEOUT
              value: 5m
            - name: OPERATOR_CONCURRENT_SCAN_JOBS_LIMIT
              value: '10'
            - name: OPERATOR_CONCURRENT_NODE_COLLECTOR_LIMIT
              value: '1'
            - name: OPERATOR_SCAN_JOB_RETRY_AFTER
              value: 30s
            - name: OPERATOR_BATCH_DELETE_LIMIT
              value: '10'
            - name: OPERATOR_BATCH_DELETE_DELAY
              value: 10s
            - name: OPERATOR_METRICS_BIND_ADDRESS
              value: ':8080'
            - name: OPERATOR_METRICS_FINDINGS_ENABLED
              value: 'true'
            - name: OPERATOR_METRICS_VULN_ID_ENABLED
              value: 'false'
            - name: OPERATOR_HEALTH_PROBE_BIND_ADDRESS
              value: ':9090'
            - name: OPERATOR_VULNERABILITY_SCANNER_ENABLED
              value: 'true'
            - name: OPERATOR_VULNERABILITY_SCANNER_SCAN_ONLY_CURRENT_REVISIONS
              value: 'true'
            - name: OPERATOR_SCANNER_REPORT_TTL
              value: 24h
            - name: OPERATOR_CONFIG_AUDIT_SCANNER_ENABLED
              value: 'true'
            - name: OPERATOR_RBAC_ASSESSMENT_SCANNER_ENABLED
              value: 'true'
            - name: OPERATOR_INFRA_ASSESSMENT_SCANNER_ENABLED
              value: 'true'
            - name: OPERATOR_CONFIG_AUDIT_SCANNER_SCAN_ONLY_CURRENT_REVISIONS
              value: 'true'
            - name: OPERATOR_EXPOSED_SECRET_SCANNER_ENABLED
              value: 'true'
            - name: OPERATOR_METRICS_EXPOSED_SECRET_INFO_ENABLED
              value: 'false'
            - name: OPERATOR_WEBHOOK_BROADCAST_URL
            - name: OPERATOR_WEBHOOK_BROADCAST_TIMEOUT
              value: 30s
            - name: OPERATOR_PRIVATE_REGISTRY_SCAN_SECRETS_NAMES
              value: '{}'
            - name: OPERATOR_ACCESS_GLOBAL_SECRETS_SERVICE_ACCOUNTS
              value: 'true'
            - name: OPERATOR_BUILT_IN_TRIVY_SERVER
              value: 'false'
            - name: TRIVY_SERVER_HEALTH_CHECK_CACHE_EXPIRATION
              value: 10h
            - name: OPERATOR_MERGE_RBAC_FINDING_WITH_CONFIG_AUDIT
              value: 'false'
          resources: {}
          livenessProbe:
            httpGet:
              path: /healthz/
              port: probes
              scheme: HTTP
            initialDelaySeconds: 5
            timeoutSeconds: 1
            periodSeconds: 10
            successThreshold: 1
            failureThreshold: 10
          readinessProbe:
            httpGet:
              path: /readyz/
              port: probes
              scheme: HTTP
            initialDelaySeconds: 5
            timeoutSeconds: 1
            periodSeconds: 10
            successThreshold: 1
            failureThreshold: 3
          terminationMessagePath: /dev/termination-log
          terminationMessagePolicy: File
          imagePullPolicy: IfNotPresent
          securityContext:
            capabilities:
              drop:
                - ALL
            privileged: true
            readOnlyRootFilesystem: true
            allowPrivilegeEscalation: true
      restartPolicy: Always
      terminationGracePeriodSeconds: 30
      dnsPolicy: ClusterFirst
      serviceAccountName: trivy-operator
      serviceAccount: trivy-operator
      automountServiceAccountToken: true
      securityContext: {}
      schedulerName: default-scheduler
  strategy:
    type: Recreate
  revisionHistoryLimit: 10
  progressDeadlineSeconds: 600

```

**Additional context**

### LINKED ISSUE #7381: CRD store precedes store implemented by extension after extension install
**Describe the bug**

If user has CRD installed in the cluster (in this case Trivy operator) and then installs an extension implementing and registering the same api/store (in this case the Trivy Operator Lens extension), the extension will get the CRD store implementation instead of it's own. 

Both stores exist but the first one in list is used (=CRD store). This only happens after extension install. If lens is restarted after the extension is installed, I believe the CRD store is not created.

Related to #7364 which was hopefully fixed by https://github.com/lensapp/oci-image-lens-extension/pull/837

**To Reproduce**
1. Install Trivy Operator to cluster
2. Wait for CRDs to be created
3. Install Trivy Operator Lens Extension
4. Navigate to Trivy Operator Extension listing page and open some vulnerability details
5. Methods implemented by the extension are missing in KubeObjects and error is thrown about `getImageRef`.

**Expected behavior**
Listing view and details open and extension is able to call the methods it implements for the store/objects.

**Environment (please complete the following information):**
- Lens Version: Local build from master
- OS: MacOs

**Logs:**
When you run the application executable from command line you will see some logging output. Please paste them here:
```
Error stack:
TypeError: e.getImageRef is not a function
```

### PR DESCRIPTION
**Pull request by Nokel81**
- Switch to auto injectable registrations

Signed-off-by: Sebastian Malton <sebastian@malton.name>

fixes https://github.com/lensapp/lens/issues/7381
fixes https://github.com/lensapp/lens/issues/7364

### CODE REVIEW

**Comment by Iku-turso** [comment]
Name of PR doesn't explain what exactly needs to be fixed.

File: packages/core/src/common/k8s-api/__tests__/api-manager.test.ts [hunk 1]
```diff
@@ -117,4 +120,58 @@ describe("ApiManager", () => {
       });
     });
   });
+
+  describe("given than a CRD has a default KubeApi registered for it", () => {
```
Comment (Iku-turso) [inline packages/core/src/common/k8s-api/__tests__/api-manager.test.ts:125]: The technical unit tests for this code are welcome, but they reveal a multitude of design problems (listed elsewhere in PR). I remind myself that unit tests like this should not couple us to those problems, but instead enable us to fix the problems without making unreasonable changes in tests.

Could this be tested behaviourally or in larger scope to enable aforementioned refactorability?

File: packages/core/src/common/k8s-api/__tests__/api-manager.test.ts [hunk 2]
```diff
@@ -117,4 +120,58 @@ describe("ApiManager", () => {
...
+          id: `default-kube-api-for-custom-resource-definition-${apiBase}`,
+          instantiate: (di) => {
+            const objectConstructor = class extends KubeObject {
+              static readonly kind = "VulnerabilityReport";
+              static readonly namespaced = true;
+              static readonly apiBase = apiBase;
+            };
+
+            return Object.assign(
+              new KubeApi({
```
Comment (Iku-turso) [inline packages/core/src/common/k8s-api/__tests__/api-manager.test.ts:140]: Coupled design problem: god class.

File: packages/core/src/common/k8s-api/__tests__/api-manager.test.ts [hunk 3]
```diff
@@ -117,4 +120,58 @@ describe("ApiManager", () => {
...
+                logger: di.inject(loggerInjectable),
+                maybeKubeApi: di.inject(maybeKubeApiInjectable),
+              }, { objectConstructor }),
```
Comment (Iku-turso) [inline packages/core/src/common/k8s-api/__tests__/api-manager.test.ts:143]: Coupled design problem: exotic pattern of injection of an `objectConstructor` instead of a factory function, or, if possible, an instance.

File: packages/core/src/common/k8s-api/__tests__/api-manager.test.ts [hunk 4]
```diff
@@ -117,4 +120,58 @@ describe("ApiManager", () => {
...
+
+    beforeEach(() => {
+      runInAction(() => {
+        di.register(getInjectable({
+          id: `default-kube-api-for-custom-resource-definition-${apiBase}`,
+          instantiate: (di) => {
+            const objectConstructor = class extends KubeObject {
+              static readonly kind = "VulnerabilityReport";
+              static readonly namespaced = true;
+              static readonly apiBase = apiBase;
```
Comment (Iku-turso) [inline packages/core/src/common/k8s-api/__tests__/api-manager.test.ts:136]: Coupled design problem: exotic and overly complicated usage of static properties.

File: packages/core/src/common/utils/registrator-helper.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+ * Copyright (c) OpenLens Authors. All rights reserved.
+ * Licensed under MIT License. See LICENSE in root directory for more information.
+ */
+
+import { iter } from "@k8slens/utilities";
+import type { DiContainerForInjection, Injectable } from "@ogre-tools/injectable";
+
+// Register new injectables and deregister removed injectables by id
+
+export const injectableDifferencingRegistratorWith = (di: DiContainerForInjection) => (
```
Comment (Iku-turso) [inline packages/core/src/common/utils/registrator-helper.ts:11]: This makes me think if `computedInjectMany` could provide also `arrivingInstances` and `departingInstances`, along with the normal `allInstances`?

File: packages/core/src/extensions/extension-loader/extension/extension.injectable.ts [hunk 6]
```diff
@@ -46,21 +36,8 @@ const extensionInjectable = getInjectable({
...
-
-                    if (idsToRemove.length > 0) {
-                      childDi.deregister(...idsToInjectables(idsToRemove, previousInjectables));
-                    }
-
-                    if (idsToAdd.length > 0) {
-                      childDi.register(...idsToInjectables(idsToAdd, currentInjectables));
-                    }
-                  }, {
+                  injectableDifferencingRegistrator,
```
Comment (Iku-turso) [inline packages/core/src/extensions/extension-loader/extension/extension.injectable.ts:41]: This (and code before it) means we dynamically register and deregister `injectables` from extension's `injectables` that get reactively observed using `computedInjectMany`. This feels complicated, let's be extra-cautious about it.

File: packages/core/src/common/utils/registrator-helper.ts (hunk 5 above)
Comment (Iku-turso) [inline packages/core/src/common/utils/registrator-helper.ts:11]: Integral utility needs technical unit tests.

File: packages/core/src/common/utils/registrator-helper.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+  (rawCurrent: Injectable<any, any, any>[], rawPrevious: Injectable<any, any, any>[] = []) => {
+    const current = new Map(rawCurrent.map(inj => [inj.id, inj]));
+    const previous = new Map(rawPrevious.map(inj => [inj.id, inj]));
+    const toAdd = iter.chain(current.entries())
+      .filter(([id]) => previous.has(id))
+      .collect(entries => new Map(entries));
+    const toRemove = iter.chain(previous.entries())
+      .filter(([id]) => !current.has(id))
+      .collect(entries => new Map(entries));
```
Comment (Iku-turso) [inline packages/core/src/common/utils/registrator-helper.ts]: FYI: Lodash has functions for set-theory that could make this sort of stuff a tad simpler, eg. `union`, `intersection` and `difference`.

File: packages/core/src/common/utils/registrator-helper.ts [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+import { iter } from "@k8slens/utilities";
+import type { DiContainerForInjection, Injectable } from "@ogre-tools/injectable";
+
+// Register new injectables and deregister removed injectables by id
+
+export const injectableDifferencingRegistratorWith = (di: DiContainerForInjection) => (
+  (rawCurrent: Injectable<any, any, any>[], rawPrevious: Injectable<any, any, any>[] = []) => {
+    const current = new Map(rawCurrent.map(inj => [inj.id, inj]));
+    const previous = new Map(rawPrevious.map(inj => [inj.id, inj]));
+    const toAdd = iter.chain(current.entries())
```
Comment (Iku-turso) [inline packages/core/src/common/utils/registrator-helper.ts:15]: Naggity nag nag: `iter` adds noise and is the wrong tool for the job here. Generators serve no tangible advantage here over good ol' map and filter.

File: packages/core/src/renderer/components/+config-horizontal-pod-autoscalers/hpa-details.test.tsx [hunk 9]
```diff
@@ -41,6 +46,17 @@ describe("<HpaDetails/>", () => {
   beforeEach(() => {
     const di = getDiForUnitTesting();
 
+    di.override(directoryForUserDataInjectable, () => "/some-user-store-path");
+    di.override(directoryForKubeConfigsInjectable, () => "/some-kube-configs");
+    di.override(storesAndApisCanBeCreatedInjectable, () => true);
+    di.override(hostedClusterInjectable, () => new Cluster({
```
Comment (Iku-turso) [inline packages/core/src/renderer/components/+config-horizontal-pod-autoscalers/hpa-details.test.tsx:52]: Loosely related: We were talking before with the pro-team, and we discussed if it would be nice to make `di` list all the injectable-ids that are overridden. With that, we would have hint of eg. unexpected global-overrides that could affect current unit test in progress. Sound good? :)

**Review by Iku-turso — CHANGES_REQUESTED** [review]

File: packages/core/src/common/utils/registrator-helper.ts (hunk 8 above)
Comment (Nokel81) [reply packages/core/src/common/utils/registrator-helper.ts:15]: Disagree, iterators are a fundamental software concept

File: packages/core/src/common/utils/registrator-helper.ts (hunk 5 above)
Comment (Nokel81) [reply packages/core/src/common/utils/registrator-helper.ts:11]: Sure, though this code is covered by unit tests indirectly

File: packages/core/src/common/utils/registrator-helper.ts (hunk 8 above)
Comment (Iku-turso) [reply packages/core/src/common/utils/registrator-helper.ts:15]: One can list other fundamental concepts which are also wrong tool for the job.

But we can agree to disagree. Later. With beer perhaps? ;)

Resolved.

File: packages/core/src/common/utils/registrator-helper.ts (hunk 5 above)
Comment (Iku-turso) [reply packages/core/src/common/utils/registrator-helper.ts:11]: It is useful to test edge-cases, and to work as documentation for other developers using it.

Comment (Iku-turso) [reply packages/core/src/common/utils/registrator-helper.ts:11]: Still: no need to see the effort, if [this](https://github.com/lensapp/lens/pull/7384#discussion_r1144397853) makes more sense in near future.

**Review by Iku-turso — APPROVED** [review]

File: packages/core/src/common/utils/registrator-helper.ts (hunk 5 above)
Comment (Nokel81) [reply packages/core/src/common/utils/registrator-helper.ts:11]: That would be good
